### PR TITLE
core: add RestartSteps= and RestartSecMax= to dynamically increase interval between auto restarts

### PR DESCRIPTION
**Pull request by YHNdnzj**
RestartSteps= accepts a positive integer as the number of steps
to take to increase the interval between auto-restarts from
RestartSec= to RestartSecMax=, or 0 to disable it.

Closes https://github.com/systemd/systemd/issues/6129

### CODE REVIEW

**Comment by poettering** [comment]
hmm, so what about this: let's change this so that RestartSecSteps= takes a number of steps to keep exponentially growing the timeout before we reach the plateau of RestartSecMax=.

i.e. with RestartSec=100ms, RestartSecMax=100s, RestartSteps=5, we'd start with a holdoff timeout of 100ms, and increase it in 5 exponential steps until we reach 100s. It appears to me that that's the best way to express what people might configure.

i.e. calculate time to sleep as `ri >= rn ? rm : (usec_t) ((rs + pow(pow(rm - rs, 1.0 / rn), ri))` or so, with `rm` being the `RestartSecMax=` value, `rs` the being `RestartSec=`, `rn` being `RestartSteps=` and `ri` the current restart iteration.

File: src/core/service.c [hunk 1]
```diff
@@ -262,6 +265,27 @@ static void service_start_watchdog(Service *s) {
...
+        if (s->n_restarts == 0 ||
+            s->restart_steps == 0 ||
+            s->restart_usec_max == USEC_INFINITY ||
+            s->restart_usec == s->restart_usec_max)
+                return s->restart_usec;
+
+        if (s->n_restarts >= s->restart_steps)
+                return s->restart_usec_max;
+
+        unit = powl(s->restart_usec_max - s->restart_usec, 1.0L / s->restart_steps);
```
Comment (poettering) [inline src/core/service.c:294]: add an assert here that restartusec_max really is >= restart_usec. (and maybe a comment that this is enforced in service_verify()

File: src/core/service.c [hunk 2]
```diff
@@ -262,6 +265,27 @@ static void service_start_watchdog(Service *s) {
...
+
+        s->restart_usec_cur = usec_add(s->restart_usec, powl(unit, s->n_restarts));
```
Comment (poettering) [inline src/core/service.c]: maybe cast the powl() result explicitly to usec_t?

File: src/core/service.h [hunk 3]
```diff
@@ -116,6 +116,9 @@ struct Service {
         char *pid_file;
 
         usec_t restart_usec;
+        unsigned restart_steps;
+        usec_t restart_usec_max;
+        usec_t restart_usec_cur;
```
Comment (poettering) [inline src/core/service.h]: why store this btw? why no calculate this on the fly whenever we need it?

Comment (YHNdnzj) [reply src/core/service.h]: We need the value before and after `n_restarts` is updated, which will require extra handling in that way. So I just made the conclusion that it is better to store it 🤔

Comment (YHNdnzj) [reply src/core/service.h]: Also I find it useful to expose the actual time we'd be waiting through dbus. Storing it might be more efficient?

Comment (YHNdnzj) [reply src/core/service.h]: Hmm, don't mind me. I reevaluated this and now I think the other way is better

File: src/core/service.c [hunk 4]
```diff
@@ -114,6 +115,8 @@ static void service_init(Unit *u) {
         s->timeout_abort_usec = u->manager->default_timeout_abort_usec;
         s->timeout_abort_set = u->manager->default_timeout_abort_set;
         s->restart_usec = u->manager->default_restart_usec;
+        s->restart_steps = 0;
```
Comment (poettering) [inline src/core/service.c]: this is redundant. all fields are zero-initialized by default. we only need to initialize the fields that shall not be zero-initialized here.

**Comment by poettering** [comment]
looks excellent! thanks! but can you still drop the unnecessary zero-initialization?

**Comment by positron96** [comment]
Hi guys. 
I believe you've got the exponential backoff formula wrong. Yours is `rs + pow(rm - rs, ri/rn)` with addition/subtraction, but it should be `rs * pow(rm/rs, ri/rn)` with multiplication/division. Consider the example of rather sane settings of 1s to 600s in 6 steps. If the formula in code operates on microseconds, it produces values 1s, 1s, 1s, 1.1s, 11s, 600s (in microseconds: 1000001, 1000056, 1003243, 1184695, 11518200, 600000000). Note that 2nd step is only 57us longer than minimal, 3rd is 3ms longer, 4th is 184ms longer, 5th is 10s longer, and 6th is 10 minutes. Another minor issue is that the formula does not produce minimal value on 1st step, but the code solves this by returning early at 1st step anyway.
The proper formula gives values 1s, 3.5s, 13s, 46s, 166s, 600s, which seems more in alignment to what is expected.

I've plotted the values for the formulas:
image

**Comment by bluca** [comment]
Can you send a PR with the change?

**Comment by yuwata** [comment]
@positron96 Your suggestion is implemented in #28266. PTAL.

**Comment by positron96** [comment]
@yuwata  Thanks for the PR, now the formula looks good to me!

**Pull request merged**